### ISSUE TITLE
Clock drift: sign of allowed_clock_drift and logging resolution

### ISSUE THREAD

**Issue by sronsiek** [onelogin/ruby-saml#599]
Env:
OS: Sles12 SP3
Rails: 5.2.6
ruby-saml-1.12.2

In the logs of our saml authenticated app I've seen a burst of occurrences of entries such as:
`Current time is earlier than NotBefore condition (2021-06-25 08:56 < 2021-06-25 08:56)`

In a log covering almost 3 months, I see ~10 instances, all within the space of an hour.

Unfortunately the resolution of the printed timestamps does not include seconds (or ms), so we cannot see the real diffs.

The behaviour seems to match exacly what is described under the `Clock Drift` heading for this gem, however looking at
the code in `ruby-saml-1.12.2/lib/onelogin/ruby-saml/response.rb` I see:

```
now = Time.now.utc

if not_before && (now_with_drift = now + allowed_clock_drift) < not_before
          error_msg = "Current time is earlier than NotBefore condition (#{now_with_drift} < #{not_before})"
...
```

This suggests to me that a **negative** value of `:allowed_clock_drift` would need to be set in order to avoid failures incurred by small time diffs.

1. Improvement suggestion: can the timestamps in the log messages be printed down to ms accuracy?
2. Can you confirm an options setting of `:allowed_clock_drift => -1.second` would be appropriate here?

Thanks,
Stefan

**Comment by pitbulk** [comment]
1. I believe that the date format used on the error message depends on the default date format of the environment, but I agree we could force it. Do you have a PR?
2. allowed_clock_drift expects a positive number, and it will consider the number of seconds to be added. If you want to set as allowed_clock_drift 5 minutes, then set 300.

**Comment by sronsiek** [comment]
1. I believe a PR is a pull request? I don't have one, and being unfamiliar with github would prefer not to indulge in messing up your vcs ;)
2. Yes - my bad.

**Comment by johnnyshields** [comment]
Fixed in #605
